services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/SemanticDetector.java

```diff
@@ -0,0 +1,438 @@
+   * @param targetSchema the target schema we want to compare with
+   * @param name The name of the map from the parent object
+   * @param defaultValueEntry The default value of the map
+   * @throws VeniceProtocolException if there is a semantic change between the two schemas & value from @code{@param map} is non-default value.
+   */
+  public static void traverseMap(
+      Map<String, Object> map,
+      Schema currentSchema,
+      Schema targetSchema,
+      String name,
+      Object defaultValueEntry) {
+    Schema currentValueSchema = currentSchema.getValueType();
+    Schema targetValueSchema = targetSchema.getValueType();
+    String mapName = buildFieldPath(name, currentSchema.getName());
+    String nestedMapName = buildFieldPath(mapName, currentSchema.getValueType().getName());
+
+    for (Map.Entry<String, Object> entry: map.entrySet()) {
+      traverseAndValidate(
+          entry.getValue(),
+          currentValueSchema,
+          targetValueSchema,
+          buildFieldPath(nestedMapName, entry.getKey()),
+          defaultValueEntry);
+    }
+  }
+
+  /**
+   * Traverse the union to make sure the object is not using new semantic.
+   * <p>
+   *   For example,
+   *   <ul>
+   *     <li>Current schema: Union (RecordA - field1, field2, field3, RecordB - field1, field2) </li>
+   *     <li>Target schema: Union (RecordA - field1, field2, field3) </li>
+  *     </ul>
+   *     In this case, we want to check the object against the schema within the union schema.
+   *     If object - RecordA(field1, field2, field3), then we are NOT using new semantic change.
+   *     If object - RecordB(field1, field2), then we have a semantic change.
+   *     If the object is using new semantic, we return true.
+   * </p>
+   * @param object the object containing values
+   * @param currentSchema the union schema
+   * @param targetSchema the target schema we want to compare with
+   * @param name The name of the object from the parent object
+   * @param defaultValue The default value of the object
+   * @throws VeniceProtocolException if there is a semantic change between the two schemas & value from @code{@param object} is non-default value
+   * OR if the object does not match any schema within the union schemas.
+   */
+  public static void traverseUnion(
+      Object object,
+      Schema currentSchema,
+      Schema targetSchema,
+      String name,
+      Object defaultValue) {
+    List<Schema> schemasInUnion = currentSchema.getTypes();
+    Schema objectSchema = getObjectSchema(object, currentSchema);
+
+    if (objectSchema == null) {
+      throw new VeniceProtocolException(
+          String.format(
+              "Field %s: Object %s does not match any schema within union schemas %s",
+              formatFieldName(name),
+              object,
+              schemasInUnion));
+    }
+
+    Map<String, Schema> targetSchemaMap =
+        targetSchema.getTypes().stream().collect(Collectors.toMap(Schema::getName, s -> s));
```

Comment:
Just for my education, primitive avro types also have names?

Reply:
primitive type does not have name, but the `Schema` has name.  For primitive types, the name is the same with type
```java
  public void testFieldName() {
    Schema unionSchema = Schema.createUnion(Arrays.asList(stringSchema, intSchema));
    Map<String, Schema> targetSchemaMap =
        unionSchema.getTypes().stream().collect(Collectors.toMap(Schema::getName, s -> s));

    System.out.println(targetSchemaMap); // {string="string", int="int"}
  }
```